openmls/src/group/core_group/new_from_external_init.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{
+    ciphersuite::signable::Verifiable,
+    credentials::CredentialBundle,
+    framing::plaintext::MlsPlaintext,
+    group::errors::ExternalInitError,
+    key_packages::KeyPackageBundle,
+    messages::{
+        proposals::{ExternalInitProposal, Proposal},
+        public_group_state::{PublicGroupState, VerifiablePublicGroupState},
+    },
+};
+
+use super::{
+    create_commit_params::{CommitType, CreateCommitParams},
+    proposals::{ProposalStore, StagedProposal},
+    CoreGroup,
+};
+use crate::group::core_group::*;
+
+pub type ExternalInitResult = Result<
+    (
+        CoreGroup,
+        MlsPlaintext,
+        Option<Welcome>,
+        Option<KeyPackageBundle>,
+    ),
+    ExternalInitError,
+>;
+
+impl CoreGroup {
+    /// Join a group without the help of an internal member. This function
+    /// requires a `PublicGroupState`, as well as the corresponding public tree
+    /// `nodes`. After the group state is initialized, this function creates an
+    /// `ExternalInit` proposal and commits it along with the given proposals by
+    /// reference and by value.
+    ///
+    /// Returns the new `CoreGroup` object, as well as the `MlsPlaintext`
+    /// containing the commit.
+    pub fn new_from_external_init(
+        backend: &impl OpenMlsCryptoProvider,
+        framing_parameters: FramingParameters,
+        nodes_option: Option<Vec<Option<Node>>>,
+        credential_bundle: &CredentialBundle,
+        proposals_by_reference: &[MlsPlaintext],
+        proposals_by_value: &[Proposal],
+        verifiable_public_group_state: VerifiablePublicGroupState,
+    ) -> ExternalInitResult {
+        let ciphersuite = Config::ciphersuite(verifiable_public_group_state.ciphersuite())?;
+        if !Config::supported_versions().contains(&verifiable_public_group_state.version()) {
+            return Err(ExternalInitError::UnsupportedMlsVersion);
+        }
+
+        // Build the ratchet tree
+
+        // Set nodes either from the extension or from the `nodes_option`.
+        // If we got a ratchet tree extension in the welcome, we enable it for
+        // this group. Note that this is not strictly necessary. But there's
+        // currently no other mechanism to enable the extension.
+        let (nodes, enable_ratchet_tree_extension) =
+            match try_nodes_from_extensions(&verifiable_public_group_state.other_extensions())? {
```

Comment:
Sorry, but I'm having a hard time wrapping my head around what this would look like with an `and_then`. Note, that the code had to change slightly since I change the input to be a slice of nodes rather than a vector.

Reply:
No worries, just a though. No need to change this here.